File: src/queue.ts
import type { State, Word } from "./types";

function hash(text: string) {
  let value = 2166136261;
  for (const char of text)
    value = Math.imul(value ^ char.charCodeAt(0), 16777619);
  return value >>> 0;
}

function seededSample<T>(items: T[], count: number, seed: string) {
  const copy = [...items];
  let value = hash(seed);
  for (let index = copy.length - 1; index > 0; index--) {
    value = (Math.imul(value, 1664525) + 1013904223) >>> 0;
    const target = value % (index + 1);
    [copy[index], copy[target]] = [copy[target], copy[index]];
  }
  return copy.slice(0, count);
}

/**
 * Builds a deterministic daily queue.
 *
 * Due reviews are interleaved with a fixed quota sampled from five frequency
 * strata. The date is part of the seed, so reloading never changes today's set.
 */
export function buildStudyQueue(
  words: Word[],
  state: State,
  now = new Date(),
  newWordQuota = 20,
  strata = 5,
) {
  const studied = new Set(state.history.map((event) => event.wordId));
  const due = words.filter(
    (word) =>
      studied.has(word.id) &&
      !state.reviews[word.id].mastered &&
      new Date(state.reviews[word.id].due) <= now,
  );
  const unseen = words.filter((word) => !studied.has(word.id));
  const date = now.toISOString().slice(0, 10);
  const indexById = new Map(words.map((word, index) => [word.id, index]));
  const introducedToday = state.history.filter(
    (event) =>
      event.date.slice(0, 10) === date &&
      event.previousReview?.repetitions === 0,
  );
  const quotas = Array.from(
    { length: strata },
    (_, layer) =>
      Math.floor(newWordQuota / strata) +
      (layer < newWordQuota % strata ? 1 : 0),
  );
  const layers = Array.from({ length: strata }, (_, layer) => {
    const start = Math.floor((unseen.length * layer) / strata);
    const end = Math.floor((unseen.length * (layer + 1)) / strata);
    const used = introducedToday.filter((event) => {
      const index = indexById.get(event.wordId) ?? -1;
      return (
        index >= Math.floor((words.length * layer) / strata) &&
        index < Math.floor((words.length * (layer + 1)) / strata)
      );
    }).length;
    const layerWords = unseen.filter((word) => {
      const index = indexById.get(word.id) ?? -1;
      return (
        index >= Math.floor((words.length * layer) / strata) &&
        index < Math.floor((words.length * (layer + 1)) / strata)
      );
    });
    return seededSample(
      layerWords,
      Math.max(0, quotas[layer] - used),
      `${date}:${layer}`,
    );
  });
  const newWords = Array.from({ length: Math.max(...quotas, 0) }, (_, round) =>
    layers.map((layer) => layer[round]),
  )
    .flat()
    .filter(Boolean);
  const queue: Word[] = [];
  let oldIndex = 0,
    newIndex = 0;
  while (oldIndex < due.length || newIndex < newWords.length) {
    for (let count = 0; count < 2 && oldIndex < due.length; count++)
      queue.push(due[oldIndex++]);
    if (newIndex < newWords.length) queue.push(newWords[newIndex++]);
  }
  return queue;
}

File: internal/core/queue.go
package core

import (
	"hash/fnv"
	"time"
)

func seededSample(items []Word, count int, seed string) []Word {
	copyOfItems := append([]Word(nil), items...)
	hasher := fnv.New32a()
	_, _ = hasher.Write([]byte(seed))
	value := hasher.Sum32()
	for index := len(copyOfItems) - 1; index > 0; index-- {
		value = value*1664525 + 1013904223
		target := int(value % uint32(index+1))
		copyOfItems[index], copyOfItems[target] = copyOfItems[target], copyOfItems[index]
	}
	if count > len(copyOfItems) {
		count = len(copyOfItems)
	}
	if count < 0 {
		count = 0
	}
	return copyOfItems[:count]
}

// BuildStudyQueue interleaves two due reviews with one stratified new word.
func BuildStudyQueue(words []Word, state State, now time.Time, quota int) []Word {
	studied := make(map[string]bool)
	for _, event := range state.History {
		studied[event.WordID] = true
	}

	due := make([]Word, 0)
	unseen := make([]Word, 0)
	for _, word := range words {
		if !studied[word.ID] {
			unseen = append(unseen, word)
			continue
		}
		review := state.Reviews[word.ID]
		dueAt, err := time.Parse(time.RFC3339Nano, review.Due)
		if !review.Mastered && (err != nil || !dueAt.After(now)) {
			due = append(due, word)
		}
	}

	const strata = 5
	date := DateKey(now)
	indexByID := make(map[string]int, len(words))
	for index, word := range words {
		indexByID[word.ID] = index
	}
	introduced := make([]HistoryEvent, 0)
	for _, event := range state.History {
		if len(event.Date) >= 10 && event.Date[:10] == date && event.PreviousReview != nil && event.PreviousReview.Repetitions == 0 {
			introduced = append(introduced, event)
		}
	}

	layers := make([][]Word, strata)
	maxLayerQuota := 0
	for layer := 0; layer < strata; layer++ {
		layerQuota := quota / strata
		if layer < quota%strata {
			layerQuota++
		}
		if layerQuota > maxLayerQuota {
			maxLayerQuota = layerQuota
		}
		used := 0
		for _, event := range introduced {
			index := indexByID[event.WordID]
			if index >= len(words)*layer/strata && index < len(words)*(layer+1)/strata {
				used++
			}
		}
		layerWords := make([]Word, 0)
		for _, word := range unseen {
			index := indexByID[word.ID]
			if index >= len(words)*layer/strata && index < len(words)*(layer+1)/strata {
				layerWords = append(layerWords, word)
			}
		}
		layers[layer] = seededSample(layerWords, layerQuota-used, date+":"+string(rune('0'+layer)))
	}

	newWords := make([]Word, 0, quota)
	for round := 0; round < maxLayerQuota; round++ {
		for layer := 0; layer < strata; layer++ {
			if round < len(layers[layer]) {
				newWords = append(newWords, layers[layer][round])
			}
		}
	}

	queue := make([]Word, 0, len(due)+len(newWords))
	oldIndex, newIndex := 0, 0
	for oldIndex < len(due) || newIndex < len(newWords) {
		for count := 0; count < 2 && oldIndex < len(due); count++ {
			queue = append(queue, due[oldIndex])
			oldIndex++
		}
		if newIndex < len(newWords) {
			queue = append(queue, newWords[newIndex])
			newIndex++
		}
	}
	return queue
}
